Mathlib/RingTheory/Coalgebra/Equiv.lean

```diff
@@ -0,0 +1,253 @@
+
+attribute [nolint docBlame] CoalgEquiv.toCoalgHom
+attribute [nolint docBlame] CoalgEquiv.toLinearEquiv
+
+@[inherit_doc CoalgEquiv]
+notation:50 A " ≃ₗc[" R "] " B => CoalgEquiv R A B
+
+/-- `CoalgEquivClass F R A B` asserts `F` is a type of bundled coalgebra equivalences
+from `A` to `B`.  -/
+class CoalgEquivClass (F : Type*) (R A B : outParam Type*) [CommSemiring R]
+    [AddCommMonoid A] [AddCommMonoid B] [Module R A] [Module R B]
+    [CoalgebraStruct R A] [CoalgebraStruct R B] [EquivLike F A B]
+    extends CoalgHomClass F R A B, SemilinearEquivClass F (RingHom.id R) A B : Prop
+
+namespace CoalgEquivClass
+
+variable {F R A B : Type*} [CommSemiring R] [AddCommMonoid A] [AddCommMonoid B]
+  [Module R A] [Module R B] [CoalgebraStruct R A] [CoalgebraStruct R B]
+
+/-- Reinterpret an element of a type of coalgebra equivalences as a coalgebra equivalence. -/
+@[coe]
+def toCoalgEquiv [EquivLike F A B] [CoalgEquivClass F R A B] (f : F) : A ≃ₗc[R] B :=
+  { (f : A →ₗc[R] B), (f : A ≃ₗ[R] B) with }
+
+/-- Reinterpret an element of a type of coalgebra equivalences as a coalgebra equivalence. -/
+instance instCoeToCoalgEquiv
+    [EquivLike F A B] [CoalgEquivClass F R A B] : CoeHead F (A ≃ₗc[R] B) where
+  coe f := toCoalgEquiv f
+
+end CoalgEquivClass
+
+namespace CoalgEquiv
+
+variable [CommSemiring R]
+
+section
+
+variable [AddCommMonoid A] [AddCommMonoid B] [Module R A] [Module R B]
+  [CoalgebraStruct R A] [CoalgebraStruct R B]
+
+/-- The equivalence of types underlying a linear equivalence. -/
+def toEquiv : (A ≃ₗc[R] B) → A ≃ B := fun f => f.toLinearEquiv.toEquiv
+
+theorem toEquiv_injective : Function.Injective (toEquiv : (A ≃ₗc[R] B) → A ≃ B) :=
+  fun ⟨_, _, _, _⟩ ⟨_, _, _, _⟩ h =>
+    (CoalgEquiv.mk.injEq _ _ _ _ _ _ _ _).mpr
+      ⟨CoalgHom.ext (congr_fun (Equiv.mk.inj h).1), (Equiv.mk.inj h).2⟩
+
+@[simp]
+theorem toEquiv_inj {e₁ e₂ : A ≃ₗc[R] B} : e₁.toEquiv = e₂.toEquiv ↔ e₁ = e₂ :=
+  toEquiv_injective.eq_iff
+
+theorem toCoalgHom_injective : Function.Injective (toCoalgHom : (A ≃ₗc[R] B) → A →ₗc[R] B) :=
+  fun _ _ H => toEquiv_injective <| Equiv.ext <| CoalgHom.congr_fun H
+
+instance : EquivLike (A ≃ₗc[R] B) A B where
+  inv := CoalgEquiv.invFun
+  coe_injective' _ _ h _ := toCoalgHom_injective (DFunLike.coe_injective h)
+  left_inv := CoalgEquiv.left_inv
+  right_inv := CoalgEquiv.right_inv
+
+instance : FunLike (A ≃ₗc[R] B) A B where
+  coe := DFunLike.coe
+  coe_injective' := DFunLike.coe_injective
+
+instance : CoalgEquivClass (A ≃ₗc[R] B) R A B where
+  map_add := (·.map_add')
+  map_smulₛₗ := (·.map_smul')
+  counit_comp := (·.counit_comp)
+  map_comp_comul := (·.map_comp_comul)
+
+@[simp, norm_cast]
+theorem toCoalgHom_inj {e₁ e₂ : A ≃ₗc[R] B} : (↑e₁ : A →ₗc[R] B) = e₂ ↔ e₁ = e₂ :=
+  toCoalgHom_injective.eq_iff
+
+@[simp]
+theorem coe_mk {f h h₀ h₁ h₂ h₃ h₄ h₅} :
+    (⟨⟨⟨⟨f, h⟩, h₀⟩, h₁, h₂⟩, h₃, h₄, h₅⟩ : A ≃ₗc[R] B) = f := rfl
+
+theorem coe_injective : @Function.Injective (A ≃ₗc[R] B) (A → B) CoeFun.coe :=
```

Comment:
This is not the right statement; it should be about `DFunlike.coe`.

I think you can just remove this lemma entirely?